EIPS/eip-7619.md

```diff
@@ -0,0 +1,251 @@
+eip: 7619
+title: Precompile Falcon512 generic verifier
+description: Proposal to add precompiled contract that allows generic signature verifications using Falcon512 postquantum algorithm
+author: Erick Pacheco Pedraza (@eum602), Marcos Allende <mallende@lacnet.com>, Diego Lopez León <dieguitoll@gmail.com>
+discussions-to: https://ethereum-magicians.org/t/falcon-512-precompiled-generic-signature-verifier/18569
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-07-03
+---
+
+## Abstract
+
+This standard proposes the inclusion of a precompiled signature verification function. The algorithm used is Falcon-512, candidate to be standarized by NIST and quantum resistant with security level I.
+
+## Motivation
+
+The advent of quantum computing threatens blockchain protocols and networks because they utilize non-quantum resistant cryptographic algorithms. When quantum computers become robust enough to run Shor’s algorithm on a large scale, the most used asymmetric algorithms, utilized for digital signatures and message encryption, such as RSA, (EC)DSA, and (EC)DH, will be no longer secure. Quantum computers will be able to break them within a short period of time. Today, there are hundreds of billions of dollars denominated in cryptocurrencies and other digital assets that rely on blockchain ledgers as well as thousands of blockchain-based applications storing value in blockchain networks. Cryptocurrencies and blockchain-based applications require solutions that guarantee quantum resistance in order to preserve the integrity of data and assets in these public and immutable ledgers. Most proposals for quantum-resistant blockchain networks are theoretical, require large QKD networks, or propose new quantum-resistant blockchain protocols to be built from scratch. This EIP is pioneer in proposing a solution compatible with current EVM blockchain protocols. It presents a simple mechanism to add a NIST-compliant post-quantum signature to blockchain transactions, making them quantum-resistant even when ECC cryptography becomes vulnerable against attacks by quantum computers. We have developed a Solidity implementation for the on-chain verification of this signatures, which does not scale due to the required high amount of gas. This is why **this EIP is proposing a pre-compiled smart contract** that allows to verify post-quantum signatures in a scalable manner.
+
+## Specification
+
+The precompiled signature verification function runs at address `0x65`. The required inputs are:
+
+- `message` - an arbitrary nummber of bytes representing the message that was signed
+- `public key` - Falcon Public key of 897 bytes
+- `signature` - 666 bytes (max size)
+
+Those inputs are encoded according to:
+
+```
+[4 bytes method signature][32 bytes signature offset][32 bytes public key offset][32 bytes message offset][32 bytes signature length][bytes signature][32 bytes public key length][bytes public key][32 bytes message length][bytes message]
+```
+
```

Comment:
Please include a source (see [EIP-1](https://eips.ethereum.org/EIPS/eip-1#linking-to-external-resources) for permitted sources) for the Falcon-512 algorithm. Otherwise I don't think this EIP has enough detail to be implementable.

Reply:
Hi @gaudren thanks for the feedback, I have addressed all observations. Since I have added an external source for Falcon-512 I am getting an error. Waiting for next steps and fixes, if needed.

Reply:
Adding a new PR to include post quantum external source links in EIP-1 in this PR(https://github.com/ethereum/EIPs/pull/8245) @gaudren @abcoathup